src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3742,36 +3741,39 @@ size_t os::Linux::find_large_page_size(size_t large_page_size) {
     if (entry->d_type == DT_DIR &&
         sscanf(entry->d_name, "hugepages-%zukB", &page_size) == 1) {
       // The kernel is using kB, hotspot uses bytes
-      if (large_page_size == page_size * K) {
-        closedir(dir);
-        return large_page_size;
+      if (page_size * K > (size_t)Linux::page_size()) {
+          // Add each found Large Page Size to _page_sizes
+          _page_sizes.add(page_size * K);
```

Comment:
Just realized one more thing, with this code we will enable all page sizes configured even if there are no pages "allocated" for the given size. 

Is that what we want or should we read the file nr_hugepages in the given director and only add it if the size is > 0?

Reply:
I think a more complete solution is to check the nr_hugepages. Additionally, this will be required by the solution you propose above that would consider "largest possible page size with enough pages free".

Reply:
I think so too. The "largest possible page size..." could be solved anyway, by just retrying all configured pages sizes until we find one that works. But it would be much more efficient to just try the page size that actually could work.

Reply:
One also can set nr_overcommit_hugepages>0 and have a "dynamic" large page pool this way, even with nr_hugepages=0. Moreover, these settings can change during the lifetime of the VM. I would not bother adding too much logic here. Allocating huge pages may or may not fail anyway and the VM has to be prepared to deal with failure. Just my 5c.

Reply:
The "dynamic part" might make this actually being "too much logic", otherwise if feels like a pretty reasonable check. The fact that huge pages can be added during the runtime of the JVM doesn't feel like a big problem, since most large reservations are done at startup. 

But you might be right, since we have to handle the case of a failed mapping, it might not be to big of a problem trying all possible page sizes.

Reply:
I'm actually using nr_overcommit_hugepages alot (has been a tip by Per Liden) since its so convenient. As for allocation at startup, I plan on making Metaspace large-page-able again at some point in the future; that would mean larger LP allocations may happen later in VM life too.

Reply:
Ok, maybe I should try it out as well :) 

Regarding allocation at startup vs later, is the plan to make new reservations during the run or supporting uncommit of large pages. Currently if a `ReservedSpace` is special (uses large pages), uncommit is disabled  and all pages are committed up front. Is your plan to change this or will it work by adding and removing `ReservedSpace`s. I have not had time to look at the new `Metaspace` implementation in detail yet.

Reply:
As it is now in my head, using LP on Metaspace would disable on-demand uncommitting (there is a second stage release of memory unaffected by this, where ReservedSpace segments get unmapped, but that is rare due to fragmentation. Due to the large page size uncommit on demand would be much less effective anyway than with normal pages.

I am vaguely aware however of someones (yours?) experiments with "soft uncommit" - madvise(MADV_FREE) - and was planning on playing around with this too. Depending on how that plays out it may be a way to get uncommit-like behavior for large pages.

Reply:
Yes, we've done some experiments using `madvise`. Some results looked promising and others a bit surprising, but I didn't actually looked at how it would affect `HUGETLB` large pages. But yes, it might be a way to get better behavior for large pages.

Reply:
Maybe we also could take another look at the "never-remap-hugepages" rule added with JDK-8007074. I understand why stefank did that, but maybe if one added safety measures (eg before remapping making sure that we have enough huge pages in the pool with a large margin) and combined with a switch it would be safe enough.

Reply:
Yes, and maybe take another look at if +UseLargePages could mean use THP, currently THP is only used if explicitly set and I'm not sure that is true with newer Linux kernels. We even have a comment about this: 
```
    // Don't try UseTransparentHugePages since there are known
    // performance issues with it turned on. This might change in the future.
    UseTransparentHugePages = false;
```